govcd/vsphere_resource_pool.go

```diff
@@ -0,0 +1,202 @@
+	if err != nil {
+		return nil, fmt.Errorf("error getting resource pool hardware versions: %s", err)
+	}
+
+	return &retrieved, nil
+}
+
+// GetDefaultHardwareVersion retrieves the default hardware version for a given resource pool.
+// The default version is usually the highest available, but it's not guaranteed
+func (rp ResourcePool) GetDefaultHardwareVersion() (string, error) {
+
+	versions, err := rp.GetAvailableHardwareVersions()
+	if err != nil {
+		return "", err
+	}
+
+	for _, v := range versions.SupportedVersions {
+		if v.IsDefault {
+			return v.Name, nil
+		}
+	}
+	return "", fmt.Errorf("no default hardware version found for resource pool %s", rp.ResourcePool.Name)
+}
+
+// GetResourcePoolById retrieves a resource pool by its ID (Moref)
+func (vcenter VCenter) GetResourcePoolById(id string) (*ResourcePool, error) {
+	resourcePools, err := vcenter.GetAllResourcePools(nil)
+	if err != nil {
+		return nil, err
+	}
+	for _, rp := range resourcePools {
+		if rp.ResourcePool.Moref == id {
+			return rp, nil
+		}
+	}
+	return nil, fmt.Errorf("no resource pool found with ID '%s' :%s", id, ErrorEntityNotFound)
+}
+
+// GetResourcePoolByName retrieves a resource pool by name.
+// It may fail if there are several resource pools with the same name
+func (vcenter VCenter) GetResourcePoolByName(name string) (*ResourcePool, error) {
+	resourcePools, err := vcenter.GetAllResourcePools(nil)
+	if err != nil {
+		return nil, err
+	}
+	var found []*ResourcePool
+	for _, rp := range resourcePools {
+		if rp.ResourcePool.Name == name {
+			found = append(found, rp)
+		}
+	}
+	if len(found) == 0 {
+		return nil, fmt.Errorf("no resource pool found with name '%s' :%s", name, ErrorEntityNotFound)
+	}
+	if len(found) > 1 {
+		var idList []string
+		for _, f := range found {
+			idList = append(idList, f.ResourcePool.Moref)
+		}
+		return nil, fmt.Errorf("more than one resource pool was found with name %s - use resource pool ID instead - %v", name, idList)
+	}
+	return found[0], nil
+}
+
+// GetAllResourcePools retrieves all available resource pool, across all vCenters
+func (vcdClient *VCDClient) GetAllResourcePools(queryParams url.Values) ([]*ResourcePool, error) {
+
+	vcenters, err := vcdClient.GetAllVCenters(queryParams)
+	if err != nil {
+		return nil, err
+	}
+	var result []*ResourcePool
+	for _, vc := range vcenters {
+		resourcePools, err := vc.GetAllResourcePools(queryParams)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, resourcePools...)
+	}
+	return result, nil
+}
+
+// ResourcePoolsFromIds returns a slice of resource pools from a slice of resource pool IDs
+func (vcdClient *VCDClient) ResourcePoolsFromIds(resourcePoolIds []string) ([]*ResourcePool, error) {
+	if len(resourcePoolIds) == 0 {
+		return nil, nil
+	}
+
+	var result []*ResourcePool
+
+	// 1. make sure there are no duplicates in the input IDs
+	uniqueIds := make(map[string]bool)
+	for _, id := range resourcePoolIds {
+		uniqueIds[id] = true
+	}
```

Comment:
Just wondering, is there a case where we should return an error in case of a duplicate?

Reply:
Now that you make me think of it, we should. I have added a check for it, and return an error.
If users ask for 2 resource pools to be inserted and see only one, they should be told that one of them was a duplicate